mcf-browser/src/TimelineExplorer.tsx

```diff
@@ -0,0 +1,390 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React, {Component} from 'react';
+import Select, {MultiValue} from 'react-select';
+
+import {Series} from './back-end/time-series';
+import {Grouping, groupSeriesByLocations, renderTimeGraph} from './utils';
+import {LoadingSpinner} from './LoadingSpinner';
+import {getName} from './back-end/utils';
+import {PageBar} from './PageBar';
+
+const STAT_VARS_PER_PAGE = 10;
+const INITIAL_NUM_LOCATIONS = 5;
+
+interface LocationMapping {
+  [dcid: string]: string;
+}
+
+interface SelectOption {
+  value: string;
+  label: string;
+}
+
+interface TimelineExplorerPropType {
+  /**
+  * Passes the data to be plotted
+  */
+  data: Series[];
+
+  /**
+  * Indicates if uploaded files are currently being parsed.
+  */
+  loading: boolean;
+}
+
+interface TimelineExplorerStateType {
+  /** The currently selected locations to filter by */
+  locations: string[];
+
+  /** A list of possible locations to select from */
+  locationOptions: SelectOption[]
+
+  /** The key for the select component to tell it
+  * to re-render when the locations are loaded in
+  */
+  selectKey: string;
+
+  /** A mapping from a location dcid to its name */
+  locationMapping: LocationMapping;
+
+  /** The list of groups to plot, grouped by stat var */
+  statVarGroups: Series[][];
+
+  /** Tracks which page the user is currently viewing (0-indexed) */
+  page: number;
+}
+
+/** Component to display the timeline explorer */
+class TimelineExplorer extends Component<
+  TimelineExplorerPropType,
+  TimelineExplorerStateType
+> {
+  /** Constructor for class, sets initial state
+    * @param {TimelineExplorerPropType} props the props passed in by parent
+    * component
+    */
+  constructor(props: TimelineExplorerPropType) {
+    super(props);
+    this.state = {
+      locations: [],
+      locationOptions: [],
+      selectKey: 'unloaded',
+      locationMapping: {},
+      statVarGroups: [],
+      page: 0,
+    };
+  }
+
+  /**
+   * Update the state variables whenever necessary given
+   * the current value of this.props
+   */
+  updateState() {
+    const locationsPromise = this.getAllLocations();
+    locationsPromise.catch(
+        () => {
+          console.log(
+              'Error getting locations for data:',
+          );
+          console.log(this.props.data);
+        },
+    );
+    locationsPromise.then(
+        (locationOptions) => {
+          const locations = locationOptions.slice(0, INITIAL_NUM_LOCATIONS)
+              .map(
+                  (option) => option.value,
+              );
+          const selectKey = locations.join(',');
+          const locationMapping: LocationMapping = {};
+          locationOptions.map(
```

Comment:
I think this should be a `.forEach()` instead of a `.map()`

Reply:
Oh yes, thanks it's been changed